swift/extractor/SwiftDispatcher.h

```diff
@@ -0,0 +1,178 @@
+#pragma once
+
+#include "swift/extractor/trap/TrapArena.h"
+#include "swift/extractor/trap/TrapLabelStore.h"
+// autogenerated file
+#include "swift/extractor/trap/TrapClasses.h"
+#include "swift/extractor/SwiftTagTraits.h"
+#include <swift/AST/SourceFile.h>
+#include <swift/Basic/SourceManager.h>
+#include <llvm/Support/FileSystem.h>
+
+namespace codeql {
+
+namespace detail {
+
+// The following `getKindName`s are used within "TBD" TRAP entries to visually mark an AST node as
+// not properly emitted yet.
+// TODO: To be replaced with QL counterpart
+template <typename Parent, typename Kind>
+inline std::string getKindName(Kind kind) {
+  return Parent::getKindName(kind).str();
+}
+
+template <>
+inline std::string getKindName<swift::TypeBase, swift::TypeKind>(swift::TypeKind kind) {
+  switch (kind) {
+#define TYPE(CLASS, PARENT)    \
+  case swift::TypeKind::CLASS: \
+    return #CLASS;
+#include "swift/AST/TypeNodes.def"
+    default:
+      return "Unknown";
+  }
+}
+
+template <>
+std::string inline getKindName<swift::TypeRepr, swift::TypeReprKind>(swift::TypeReprKind kind) {
+  switch (kind) {
+#define TYPEREPR(CLASS, PARENT)    \
+  case swift::TypeReprKind::CLASS: \
+    return #CLASS;
+#include "swift/AST/TypeReprNodes.def"
+    default:
+      return "Unknown";
+  }
+}
+
+}  // namespace detail
+
+// The main reponsibilities of the SwiftDispatcher are as follows:
+// * redirect specific AST node emission to a corresponding visitor (statements, expressions, etc.)
+// * storing TRAP labels for emitted AST nodes (in the TrapLabelStore) to avoid re-emission
+// Since SwiftDispatcher sees all the AST nodes, it also attaches a location to every 'locatable'
+// node (AST nodes that are not types: declarations, statements, expressions, etc.).
+class SwiftDispatcher {
+ public:
+  SwiftDispatcher(const swift::SourceManager& sourceManager, TrapArena& arena, TrapOutput& trap)
```

Comment:
Please document the lifetime requirements on the arguments.

The default assumption is that function arguments only need to be valid as long as the call, but since we capture them by reference here, they actually need to outlive the constructed object as well.